File: Problem486/Solution.go
package main

func PredictTheWinner(nums []int) bool {

	return getMaxSum(nums, 0, len(nums)-1) >= 0
}

func getMaxSum(nums []int, start int, end int) int {
	if start == end {
		return nums[start]
	}
	sum_start := nums[start] - getMaxSum(nums, start+1, end)
	sum_end := nums[end] - getMaxSum(nums, start, end-1)
	if sum_start > sum_end {
		return sum_start
	}
	return sum_end
}

File: Problem486/Solution.java
public class Solution
{
    public boolean PredictTheWinner(int[] nums)
    {
        return getSum(nums, 0, nums.length - 1) >= 0;
    }

    private int getSum(int[] nums, int start, int end)
    {
        return start == end ? nums[start]
                : Math.max(nums[end] - getSum(nums, start, end - 1), nums[start] - getSum(nums, start + 1, end));
    }
}